### ISSUE TITLE
Can't create a record with only the foreign key to another table

### ISSUE THREAD

**Issue by cannikin** [prisma/prisma#2152]
## Bug description

You are unable to create a new record by using the foreign key to another table. For example if a `Post` has a `userId` column, you cannot just set the `userId` and save the record.

## How to reproduce

Given the schema shown below, create a `User` and note its ID. Let's say it is `1` in this case. Now try to create a post:

```
db.post.create({ data: { title: 'My post', body: 'Lorem ipsum...', userId: 1 } })
```

Error:

```
Invalid `prisma.post.create()` invocation:
{
  data: { 
    title: 'First post',    
    body: 'Body goes here',    
    userId: 1,   
    ~~~~~~
    +   user: {
    +     create?: UserCreateWithoutPostsInput,
    +     connect?: UserWhereUniqueInput
    +   }  
  }
}
Unknown arg `userId` in data.userId for type PostCreateInput. Did you mean `user`?
Argument user for data.user is missing.
```

## Expected behavior

The `Post` record is created and `userId` is set to `1`.

## Prisma information

```
// schema.prisma
datasource DS {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}

generator client {
  provider      = "prisma-client-js"
}

model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String?
  posts Post[]
}

model Post {
  id     Int     @id @default(autoincrement())
  title  String  @unique
  body   String
  user   User    @relation(fields: [userId], references: [id])
  userId Int
}
```

## Environment & setup

```
rob$ yarn prisma -v
yarn run v1.22.4
$ /Users/rob/Sites/redwoodjs/abbott/node_modules/.bin/prisma -v
@prisma/cli          : 2.0.0-beta.1
Current platform     : darwin
Query Engine         : prisma 2accb9c7eacdc984874eaeb63377fe705dfd3203 (at /Users/rob/Sites/redwoodjs/abbott/node_modules/@prisma/cli/query-engine-darwin)
Migration Engine     : migration-engine-cli 2accb9c7eacdc984874eaeb63377fe705dfd3203 (at /Users/rob/Sites/redwoodjs/abbott/node_modules/@prisma/cli/migration-engine-darwin)
Introspection Engine : introspection-core 2accb9c7eacdc984874eaeb63377fe705dfd3203 (at /Users/rob/Sites/redwoodjs/abbott/node_modules/@prisma/cli/introspection-engine-darwin)
```

```
rob$ node -v
v13.8.0
```

## Why this is important

With RedwoodJS we have generators that create simple CRUD pages for working the data in your database. In this case we would create a "Post" form to let you create new ones. To associate the new post to a user we provide a `userId` input. Ideally the user just enters the ID of the user and they're done:

![image](https://user-images.githubusercontent.com/300/78950011-4af30880-7a82-11ea-8d69-1678b796e132.png)

Unfortunately this behavior breaks that completely. :( We would need a lot of code to work around this (perform an extra database query to look up the user and then include that user using the `connect` syntax, I guess?). This can all be avoided by just letting users set all the fields in the database table, which most would expect to be able to do.

**Comment by Sytten** [comment]
While I agree this could be useful, it is already possible by doing:
```typescript
db.post.create({ data: { title: 'My post', body: 'Lorem ipsum...', user: { connect: { id: 1 } } } })
```
So I would say, allow your user to input the `userId` and rewrite the query to use the `connect` syntax?

**Comment by cannikin** [comment]
@Sytten Thanks for the idea! I think we're going to implement this as a workaround for now, but ideally we could just set foreign key columns like we can set any other column in the table.

**Comment by schickling** [comment]
Hi @cannikin – a quick update on this issue: We've looked into this and while from an API perspective this is quite a small change, this would mean a significant of engineering work on the query engine level which we're not quite ready to take on.

This issue is also related to some other ongoing issues (e.g. https://github.com/prisma/prisma/issues/2018) which we want to illuminate further before moving forward with this.

I hope @Sytten's workaround works for the time being.

**Comment by cannikin** [comment]
@schickling Thanks for the update! The workaround works for now but I'll be keeping my eye on this one. 😉

**Comment by eurobob** [comment]
@cannikin I'm struggling to get this working. I'm using the redwood scaffolding and trying to update a row in my database like you mentioned here, but I receive this error:

```[GraphQL error]: Message: Variable "$input" got invalid value......Field "ambassador" is not defined by type UpdateExperienceInput. Did you mean ambassadorId?```


My `experiences.sdl.js` has this abridged Input:

```
input UpdateExperienceInput {
    ambassadorId: Int
    ...
  }
```

And my abridged schema model looks like this:

```
model Experience {
  ambassadorId     Int?
  ambassador       Ambassador?       @relation(fields: [ambassadorId], references: [id])
  ...
  @@index([ambassadorId], name: "Experience_ambassadorId_Ambassador_fk")
}
```

In my edit cell my `onSave` method looks like this:

```
const onSave = (input, id) => {
    const castInput = Object.assign(input, {
      ambassador: {
        connect: {
          id: parseInt(input.ambassador),
        },
      },
    })
    updateExperience({ variables: { id, input: castInput } })
  }
```

And in my form I have assigned the value of ambassadorId to an input named "ambassador":

```
<Label
          name="ambassador"
          className={CSS.label}
          errorClassName={CSS.labelError}
        />
        <TextField
          name="ambassador"
          defaultValue={props.experience?.ambassadorId}
          className={CSS.input}
          errorClassName={CSS.inputError}
          validation={{ required: true }}
        />
        <FieldError name="ambassador" className={CSS.errorMessage} />
```

What am I getting wrong?

Thanks in advance 🙏

**Comment by cannikin** [comment]
Looks to me like your input type `UpdateExperienceInput` expects `ambassadorId` but in `castInput` you're creating a the key `ambassador` so those types are mismatching.

Rather than create the required prisma syntax in the `input` object as you're doing, we were letting the foreign key be submitted through the GraphQL but then turning it into the `connect` syntax in the service itself. So in the service file where you're creating the `Experience` instead of just doing something like `db.experience.create({ data: input })`, it would be more like:

```js
export const createExperience = ({ input }) => {
  let output = input
  output = Object.assign(output, { ambassador: { connect: { id: input.ambassadorId } } })
  delete output.ambassadorId

  return db.experience.create({ data: output })
}
```

We had some code that would do this automatically for any field that had a name ending in `Id` and we're working on integrating this into Redwood to make it transparent to developers. But you could put this in your service (or in its own file and then import into any services that need it):

```js
export const foreignKeyReplacement = (input) => {	
  let output = input	
  const foreignKeys = Object.keys(input).filter((k) => k.match(/Id$/))	

  foreignKeys.forEach((key) => {	
    const modelName = key.replace(/Id$/, '')	
    const value = input[key]	

    delete output[key]	
    output = Object.assign(output, {	
      [modelName]: { connect: { id: value } },	
    })	
  })	

  return output	
}
```

And then to use it:

```js
export const createExperience = ({ input }) => {
  return db.experience.create({ data: foreignKeyReplacement(input) })
}
```

**Comment by eurobob** [comment]
@cannikin AMAZING! Thank you so much for the fast response and clarity.

Really love redwood, it's the system I've been wanting for years but lacking the knowledge to pull it off.

Really happy to be able to be moving away from using sequelize, so really appreciate all the work, gonna help out as much as I can!

**Comment by cannikin** [comment]
@eurobob You're welcome! :)

**Comment by jalik** [comment]
I am also very concerned about this issue. I expected Prisma to treat all foreign keys as normal fields when saving. This behaviour is very questionable compared to the rest of the Prisma code which is extremely intuitive, time saving and DX friendly (thanks by the way).

For those who are interested in a more flexible (and verbose) workaround, I wrote some helper functions to "connect" relations, it works with foreign keys that don't end with "Id", unlike @cannikin 's solution (which is fine too) :
```js
function isDefined(value) {
  return value !== null && typeof value !== 'undefined';
}

function connectRelations(data, relations) {
  const d = { ...data }; // it would be better to deep clone the data

  Object.keys(relations).forEach((key) => {
    if (Object.values(relations[key]).filter(isDefined).length > 0) {
      d[key] = { connect: relations[key] };
    }
  });
  return d;
}
```

Then use them like below :
```js
// Extract foreign keys if present
const {cityId, roleId, username} = data;

// Add connect attributes for listed foreign keys if their value is defined (!=null && !=undefined)
const result = await prisma.customer.create({
  data: connectRelations(data, {
    city: { id: cityId },
    role: { id: roleId },
    user: { username },
  }),
});
```

**Comment by dpetrick** [comment]
This is now available in `@prisma/client@2.11.0-dev.30` and will be included in the next stable release (2.11). Documentation is not yet available, you may need to consult the PR for more details for now (https://github.com/prisma/prisma-engines/pull/1315).

**Comment by JulianKingman** [comment]
I'm on @prisma/client@5.11.0 and I'm getting the same error as the OP. Do I have to enable this somehow?
